NPE in XMLTextDocumentService.computeDOMAsync

spotted in log:

```
Jan 07, 2021 1:55:18 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lemminx.commons.ParentProcessWatcher.lambda$apply$0(ParentProcessWatcher.java:143)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 more
Caused by: java.lang.NullPointerException
	at org.eclipse.lemminx.XMLTextDocumentService.computeDOMAsync(XMLTextDocumentService.java:546)
	at org.eclipse.lemminx.XMLTextDocumentService.foldingRange(XMLTextDocumentService.java:316)
	... 17 more
```
I do not have steps to reproduce.
I'm using  org.eclipse.wildwebdeveloper.xml 0.9.1.202007231627

--- a/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLTextDocumentService.java
+++ b/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLTextDocumentService.java
@@ -28,6 +28,8 @@ import java.util.function.BiFunction;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
+import com.google.gson.JsonPrimitive;
+
 import org.eclipse.lemminx.client.ExtendedClientCapabilities;
 import org.eclipse.lemminx.client.LimitExceededWarner;
 import org.eclipse.lemminx.client.LimitFeature;
@@ -93,8 +95,6 @@ import org.eclipse.lsp4j.jsonrpc.CancelChecker;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
 import org.eclipse.lsp4j.jsonrpc.validation.NonNull;
 import org.eclipse.lsp4j.services.TextDocumentService;
-
-import com.google.gson.JsonPrimitive;
 
 /**
  * XML text document service.
@@ -256,6 +256,9 @@ public class XMLTextDocumentService implements TextDocumentService {
 			DocumentSymbolParams params) {
 
 		TextDocument document = getDocument(params.getTextDocument().getUri());
+		if (document == null) {
+			return CompletableFuture.completedFuture(null);
+		}
 		XMLSymbolSettings symbolSettings = sharedSettings.getSymbolSettings();
 
 		if (!symbolSettings.isEnabled() || symbolSettings.isExcluded(document.getUri())) {
@@ -302,6 +305,9 @@ public class XMLTextDocumentService implements TextDocumentService {
 		return computeAsync((cancelChecker) -> {
 			String uri = params.getTextDocument().getUri();
 			TextDocument document = getDocument(uri);
+			if (document == null) {
+				return null;
+			}
 			CompositeSettings settings = new CompositeSettings(getSharedSettings(), params.getOptions());
 			return getXMLLanguageService().format(document, null, settings);
 		});
@@ -312,6 +318,9 @@ public class XMLTextDocumentService implements TextDocumentService {
 		return computeAsync((cancelChecker) -> {
 			String uri = params.getTextDocument().getUri();
 			TextDocument document = getDocument(uri);
+			if (document == null) {
+				return null;
+			}
 			CompositeSettings settings = new CompositeSettings(getSharedSettings(), params.getOptions());
 			return getXMLLanguageService().format(document, params.getRange(), settings);
 		});
@@ -595,7 +604,11 @@ public class XMLTextDocumentService implements TextDocumentService {
 	 */
 	public <R> CompletableFuture<R> computeDOMAsync(TextDocumentIdentifier documentIdentifier,
 			BiFunction<CancelChecker, DOMDocument, R> code) {
-		return computeModelAsync(getDocument(documentIdentifier.getUri()).getModel(), code);
+		ModelTextDocument<DOMDocument> document = getDocument(documentIdentifier.getUri());
+		if (document != null) {
+			return computeModelAsync(document.getModel(), code);
+		}
+		return CompletableFuture.completedFuture(null);
 	}
 
 	private static <R, M> CompletableFuture<R> computeModelAsync(CompletableFuture<M> loadModel,

--- a/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/commons/OutOfOrderRequestsTest.java
+++ b/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/commons/OutOfOrderRequestsTest.java
@@ -1,0 +1,186 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.commons;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+import org.eclipse.lemminx.XMLLanguageServer;
+import org.eclipse.lemminx.XMLTextDocumentService;
+import org.eclipse.lsp4j.CodeLensParams;
+import org.eclipse.lsp4j.CompletionParams;
+import org.eclipse.lsp4j.DefinitionParams;
+import org.eclipse.lsp4j.DocumentFormattingParams;
+import org.eclipse.lsp4j.DocumentHighlightParams;
+import org.eclipse.lsp4j.DocumentLinkParams;
+import org.eclipse.lsp4j.DocumentRangeFormattingParams;
+import org.eclipse.lsp4j.DocumentSymbolParams;
+import org.eclipse.lsp4j.FoldingRangeRequestParams;
+import org.eclipse.lsp4j.FormattingOptions;
+import org.eclipse.lsp4j.HoverParams;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.ReferenceContext;
+import org.eclipse.lsp4j.ReferenceParams;
+import org.eclipse.lsp4j.RenameParams;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.eclipse.lsp4j.TypeDefinitionParams;
+import org.eclipse.lsp4j.services.TextDocumentService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests to ensure that null is returned on requests for a document that happen before the document is open
+ */
+public class OutOfOrderRequestsTest {
+
+	private static final TextDocumentIdentifier UNOPENED_DOC = new TextDocumentIdentifier("file:///file.xml");
+
+	private TextDocumentService ls = null;
+
+	@BeforeEach
+	public void setupLS() {
+		ls = new XMLTextDocumentService(new XMLLanguageServer());
+	}
+
+	@Test
+	public void completion() {
+		CompletionParams params = new CompletionParams(UNOPENED_DOC, new Position());
+		ls.completion(params) //
+		.thenAccept((completion) -> {
+			assertNull(completion);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void hover() {
+		HoverParams params = new HoverParams(UNOPENED_DOC, new Position());
+		ls.hover(params) //
+		.thenAccept((hover) -> {
+			assertNull(hover);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void documentHighlight() {
+		DocumentHighlightParams params = new DocumentHighlightParams(UNOPENED_DOC, new Position());
+		ls.documentHighlight(params) //
+		.thenAccept((documentHighlight) -> {
+			assertNull(documentHighlight);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void documentSymbol() {
+		DocumentSymbolParams params = new DocumentSymbolParams(UNOPENED_DOC);
+		ls.documentSymbol(params) //
+		.thenAccept((documentSymbol) -> {
+			assertNull(documentSymbol);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void rename() {
+		RenameParams params = new RenameParams(UNOPENED_DOC, new Position(), "");
+		ls.rename(params) //
+		.thenAccept((rename) -> {
+			assertNull(rename);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void foldingRange() {
+		FoldingRangeRequestParams params = new FoldingRangeRequestParams(UNOPENED_DOC);
+		ls.foldingRange(params) //
+		.thenAccept((foldingRange) -> {
+			assertNull(foldingRange);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void documentLink() {
+		DocumentLinkParams params = new DocumentLinkParams(UNOPENED_DOC);
+		ls.documentLink(params) //
+		.thenAccept((documentLink) -> {
+			assertNull(documentLink);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void definition() {
+		DefinitionParams params = new DefinitionParams(UNOPENED_DOC, new Position());
+		ls.definition(params) //
+		.thenAccept((definition) -> {
+			assertNull(definition);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void typeDefinition() {
+		TypeDefinitionParams params = new TypeDefinitionParams(UNOPENED_DOC, new Position());
+		ls.typeDefinition(params) //
+		.thenAccept((typeDefinition) -> {
+			assertNull(typeDefinition);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void references() {
+		ReferenceParams params = new ReferenceParams(UNOPENED_DOC, new Position(), new ReferenceContext());
+		ls.references(params) //
+		.thenAccept((references) -> {
+			assertNull(references);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void codeLens() {
+		CodeLensParams params = new CodeLensParams(UNOPENED_DOC);
+		ls.codeLens(params) //
+		.thenAccept((codeLens) -> {
+			// Code Lens is disabled by default, so it will return an empty list instead of null
+			assertNotNull(codeLens);
+			assertEquals(0, codeLens.size(), "Code lens is disabled, so it should return an empty list");
+		}) //
+		.join();
+	}
+
+	@Test
+	public void formatting() {
+		DocumentFormattingParams params = new DocumentFormattingParams(UNOPENED_DOC, new FormattingOptions());
+		ls.formatting(params) //
+		.thenAccept((formatting) -> {
+			assertNull(formatting);
+		}) //
+		.join();
+	}
+
+	@Test
+	public void rangeFormatting() {
+		DocumentRangeFormattingParams params = new DocumentRangeFormattingParams(UNOPENED_DOC, new FormattingOptions(), new Range());
+		ls.rangeFormatting(params) //
+		.thenAccept((formatting) -> {
+			assertNull(formatting);
+		}) //
+		.join();
+	}
+
+}

